### PR TITLE
Reduce use of stack based buffers for string operations

### PR DESCRIPTION
**Pull request by dirkmueller**
strprintf() allocates the right memory size for us, so we don't need static buffers and strdup()
    
Also make use of snprintf() to avoid overflows.
    
Last but not least free(NULL) is safe, so remove redundant if() checks.

### CODE REVIEW

**Comment by wfeldt** [comment]
Is that the solution for https://bugzilla.suse.com/show_bug.cgi?id=1180792?

**Comment by dirkmueller** [comment]
@wfeldt see comment in bugzilla. I can't really reproduce the issue in test mode for some reason.. I can't figure out where to get the directory /lib/modules/$kernelv/initrd that it seems to read in the installation sys into my testmode setup. 

so these changes are just from manual code review. there are piles of potentially overflowable buffers, really hard to decide where to start.. :(

**Comment by wfeldt** [comment]
I wasn't sure which was first. :-)

I'll have a look.

**Comment by wfeldt** [comment]
While code might not look ideally, I don't think it's wrong. The reported problem gets fixed with https://github.com/openSUSE/linuxrc/pull/241.

**Comment by wfeldt** [comment]
I spoke too soon. There is a bug. :-(

**Comment by dirkmueller** [comment]
@wfeldt wow, indeed, this looks like an overflow. still I think this code is a worthwhile cleanup (especially the removal of the temporary 15 bytes stack buffer in read_module_list)

**Comment by wfeldt** [comment]
I believe I've now covered everything in https://github.com/openSUSE/linuxrc/pull/241 but for the WiFi issue you reported.

I will commit the bug fixes first and then revisit this pr for the cleanup part. One hunk is already covered in my pr; the rest I'll have to check.

**Comment by dirkmueller** [comment]
@wfeldt so I realized that there is already a function for this, strprintf() from utils, so for consistency I remove the asprintf stuff again.

File: module.c [hunk 1]
```diff
@@ -267,26 +266,19 @@ void mod_update_list()
...
   while((de = readdir(d))) {
-    i = strlen(de->d_name);
-    if(
-      i < (int) sizeof buf &&
-      mod_has_module_ext(de->d_name, i, &ext_pos)
-    ) {
-      strcpy(buf, de->d_name);
-      buf[ext_pos] = 0;
+    if(mod_has_module_ext(de->d_name, strlen(de->d_name), &ext_pos)) {
+      int found = 0;
```
Comment (wfeldt) [inline module.c:277]: This is not correct as it edits out that `buf[ext_pos] = 0`. This was actually the reason for that duplicate buffer.

**Comment by wfeldt** [comment]
I've taken your code and extended it to remove all static string buffers and avoid `snprintf` completely: https://github.com/openSUSE/linuxrc/pull/244.

**Comment by dirkmueller** [comment]
thanks!

**Pull request closed without merging**